gmso/formats/mol2.py

```diff
@@ -0,0 +1,177 @@
+"""Convert to and from a TRIPOS mol2 file."""
+# TODO add sources of mol2 files
+import warnings
+from pathlib import Path
+
+import unyt as u
+
+from gmso import Atom, Bond, Box, Topology
+from gmso.core.element import element_by_name, element_by_symbol
+from gmso.formats.formats_registry import loads_as
+
+
+@loads_as(".mol2")
+def from_mol2(filename, site_type="atom"):
+    """Read in a TRIPOS mol2 file format into a gmso topology object.
+
+    Creates a Topology from a mol2 file structure. This will read in the
+    topological structure (sites, bonds, and box) information into gmso.
+    Note that parameterized information can be found in these objects, but
+    will not be converted to the Topology.
+
+    Parameters
+    ----------
+    filename : string
+        path to the file where the mol2 file is stored.
+    site_type : string ('atom' or 'lj'), default='atom'
+        tells the reader to consider the elements saved in the mol2 file, and
+        if the type is 'lj', to not try to identify the element of the site,
+        instead saving the site name.
+
+    Returns
+    -------
+    top : gmso.Topology
+
+    Notes
+    -----
+    It may be common to want to create an mBuild compound from a mol2 file. This is possible
+    by installing [mBuild](https://mbuild.mosdef.org/en/stable/index.html)
+    and converting using the following python code:
+
+        >>> import gmso.Topology
+        >>> from gmso.external.convert_mbuild import to_mbuild
+        >>> top = Topology.load('myfile.mol2')
+        >>> mbuild_compound = to_mbuild(top)
+    """
+    msg = "Provided path to file that does not exist"
+    path = Path(filename)
+    if not path.exists():
+        raise OSError(msg)
+    # Initialize topology
+    topology = Topology(name=path.stem)
+    # save the name from the filename
+    f = open(path, "r")
+    line = f.readline()
+    while f:
+        # check for header character in line
+        if line.startswith("@<TRIPOS>"):
+            # if header character in line, send to a function that will direct it properly
+            line = parse_record_type_indicator(f, line, topology, site_type)
```

Comment:
https://github.com/mosdef-hub/gmso/pull/562#discussion_r687979852

Making sure I understand this correctly, sometimes you might get the splitted line (a list) as your input. Is that correct?

Reply:
Good question. I have yet to see a mol2 file that is built with bonds before atoms. It would certainly make things difficult to build in gmso since you need to create bonds by having sites. If we wanted to support that, we would have to store the bond information somewhere separately until the site information was read in, and then go back and create the bonds. But it seems to me like a very unusual case since I've never seen it, although it doesn't explicitly specify that ordering in some of the documentation I've seen.

Reply:
In response to the #562 comment:
I think you are correct. The function `parse_record_type_indicator` is going to need the splitted line as its input so it can decide where to send the next section of information. As of the current implementation, I continue trying to read a line until the end of the section is marked by a new line, a blank line, or a new rti. Which means that we don't care if there is a space or not at the end of the line. But if there is an rti right at the end of a section to indicate a new section, I need to keep passing that line info back into `from_mol2` to determine when the new section begins. So that's the motivation behind passing that back and forth, that way extra spaces at the end of sections are handled.